File: Chapter12/C#/HostEmulator/Workers/ConnectionListener.cs
using System;
using System.Net;
using System.Net.Sockets;

namespace HostEmulator
{
  /// <summary>
  /// ConnectionListener waits for client connections. When
  /// a connection comes in, the class fires an event
  /// </summary>
  public class ConnectionListener
  {
    TcpListener listener;

    public void Listen()
    {
      IPAddress ipAddress = Dns.Resolve("localhost").AddressList[0];
      listener = new TcpListener(ipAddress, 8020); // listen on port 8020
      listener.Start();

      while (true) 
      {
        // wait until a client connects
        if (!listener.Pending() ) 
        {
          FireIdle();
          continue;
        }
          
        // process the client
        Socket socket = listener.AcceptSocket();
  			
        FireClientConnected(socket);
      }
    }

    public delegate void ClientConnectHandler(Socket theSocket);
    public event ClientConnectHandler OnClientConnect;
    public void FireClientConnected(Socket theSocket)
    {
      if (OnClientConnect != null)
        OnClientConnect(theSocket);
    }

    public delegate void IdleHandler();
    public event IdleHandler OnIdle;
    public void FireIdle()
    {
      if (OnIdle != null)
        OnIdle();
    }
  }
}

File: Chapter12/VB/HostEmulator/Workers/ConnectionListener.vb
Imports System.Net
Imports System.Net.Sockets

Public Class ConnectionListener
  Dim _listener As TcpListener

  Public Sub Listen()
    Dim ipAddress As IPAddress = Dns.Resolve("localhost").AddressList(0)
    _listener = New TcpListener(ipAddress, 8020) ' listen on port 8020
    _listener.Start()

    While True
      ' wait until a client connects
      If Not _listener.Pending() Then
        FireIdle()
      Else
        ' process the client
        Dim socket As Socket = _listener.AcceptSocket()
        FireClientConnected(socket)
      End If
    End While
  End Sub

  Public Event OnClientConnect(ByVal theSocket As Socket)
  Public Sub FireClientConnected(ByVal theSocket As Socket)
    RaiseEvent OnClientConnect(theSocket)
  End Sub

  Public Event OnIdle()
  Public Sub FireIdle()
    RaiseEvent OnIdle()
  End Sub
End Class
